### ISSUE TITLE
Change of behaviour on binary matching when value is inlined

### ISSUE THREAD

**Issue by josevalim** [erlang/otp#7469]
**Describe the bug**

Take this Erlang module:

```erlang
-module(bin_match_bug).
-export([parse/1,run/0]).

parse(<<2:2/integer,Rest:62/bitstring>>) ->
    parse_raw(<<2:2/integer>>, Rest).

parse_raw(Variant, Rest) ->
    VariantSize = bit_size(Variant),
    % VariantSize = 2,
    ClockHiSize = 8 - VariantSize,
    ClockSize = 8 + ClockHiSize,
    case Rest of
        <<ClockHi:ClockHiSize/bitstring,
          ClockLo:8/bitstring,
          _:48/bitstring>> ->
            case
                <<ClockHi:ClockHiSize/bitstring,
                  ClockLo:8/bitstring>>
            of
                <<Clock:ClockSize/integer-unsigned>> ->
                    {ok, Clock};
                Bin1 ->
                    {error1, Bin1}
            end;
        Bin2 ->
            {error2, Bin2}
    end.

run() ->
    Bin = <<184,188,2,66,172,19,0,3>>,
    {ok, 14524} = bin_match_bug:parse(Bin).
```

And now run it as `erlc bin_match_bug.erl && erl -s bin_match_bug run`. The code fails but it should work (it works on Erlang/OTP 25).

Interestingly, if you replace `VariantSize = bit_size(Variant)` by `VariantSize = 2`, which is the exact same value, it works as before.

**Affected versions**

26.0,  26.0.1,  26.0.2.

**Additional context**

Initially reported here: https://github.com/bitwalker/uniq/issues/13.

**Comment by alappe** [comment]
It works for me:

```
% erlc bin_match_bug.erl && erl --disable-jit -s bin_match_bug run
Erlang/OTP 26 [erts-14.0] [source] [64-bit] [smp:10:10] [ds:10:10:10] [async-threads:1] [jit]

Runtime terminating during boot ({{badmatch,{error1,}},[{bin_match_bug,run,0,[{_},{_}]},{init,start_em,1,[]},{init,do_boot,3,[]}]})

Crash dump is being written to: erl_crash.dump...done
```

On an Apple M1 with 26.0.2 installed via asdf.

**Comment by jhogberg** [comment]
Ah, it must be specific to the ARM JIT then, thank you. :)

**Comment by yordis** [comment]
We are facing the same issue,

Intel-based are working (my computer and CI computers), while other devs are in ARM and it is causing a problem with https://github.com/bitwalker/uniq/issues/13 Uniq package

**Comment by bitwalker** [comment]
@yordis The 0.6.1 release of `uniq` should workaround this issue for the time being, so at least you don't have to wait for the OTP patch

**Comment by yordis** [comment]
@bitwalker much love! Seriously appreciate your support! 💜

**Comment by josevalim** [comment]
This is fixed on 26.1.
